elytron/src/main/java/org/wildfly/extension/elytron/JaasRealmDefinition.java

```diff
@@ -0,0 +1,179 @@
+            .setRestartAllServices()
+            .build();
+
+    static final SimpleAttributeDefinition PATH =
+            new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.PATH, FileAttributeDefinitions.PATH)
+                    .setAttributeGroup(ElytronDescriptionConstants.FILE)
+                    .setRequired(false)
+                    .setRestartAllServices()
+                    .build();
+
+    static final SimpleAttributeDefinition RELATIVE_TO = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.RELATIVE_TO, ModelType.STRING, true)
+            .setMinSize(1)
+            .setRequired(false)
+            .setAttributeGroup(ElytronDescriptionConstants.FILE)
+            .setRequires(ElytronDescriptionConstants.PATH)
+            .setRestartAllServices()
+            .build();
+
+    static final SimpleAttributeDefinition MODULE = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.MODULE, ModelType.STRING, false)
+            .setRequired(false)
+            .setRestartAllServices()
+            .build();
+
+    private static final SimpleAttributeDefinition CALLBACK_HANDLER = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.CALLBACK_HANDLER, ModelType.STRING, true)
+            .setRequired(false)
+            .setRestartAllServices()
+            .build();
+
+    static final AttributeDefinition[] ATTRIBUTES = new AttributeDefinition[]{ENTRY, PATH, RELATIVE_TO, MODULE, CALLBACK_HANDLER};
+
+    private static final AbstractAddStepHandler ADD = new JaasRealmDefinition.RealmAddHandler();
+    private static final OperationStepHandler REMOVE = new TrivialCapabilityServiceRemoveHandler(ADD, SECURITY_REALM_RUNTIME_CAPABILITY);
+
+    JaasRealmDefinition() {
+        super(new Parameters(PathElement.pathElement(ElytronDescriptionConstants.JAAS_REALM), ElytronExtension.getResourceDescriptionResolver(ElytronDescriptionConstants.JAAS_REALM))
+                .setAddHandler(ADD)
+                .setRemoveHandler(REMOVE)
+                .setAddRestartLevel(OperationEntry.Flag.RESTART_RESOURCE_SERVICES)
+                .setRemoveRestartLevel(OperationEntry.Flag.RESTART_RESOURCE_SERVICES)
+                .setCapabilities(SECURITY_REALM_RUNTIME_CAPABILITY));
+    }
+
+    @Override
+    public void registerAttributes(ManagementResourceRegistration resourceRegistration) {
+        AbstractWriteAttributeHandler write = new ElytronReloadRequiredWriteAttributeHandler(ATTRIBUTES);
+        for (AttributeDefinition current : ATTRIBUTES) {
+            resourceRegistration.registerReadWriteAttribute(current, null, write);
+        }
+    }
+
+    private static class RealmAddHandler extends BaseAddHandler {
+
+        private RealmAddHandler() {
+            super(SECURITY_REALM_RUNTIME_CAPABILITY, ATTRIBUTES);
+        }
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
```

Comment:
```
[standalone@embedded /] /subsystem=elytron/jaas-realm=a:add(entry=e1)
{
    "outcome" => "failed",
    "failure-description" => {"WFLYCTL0080: Failed services" => {"org.wildfly.security.security-realm.a" => "Failed to start service
    Caused by: java.lang.NullPointerException"}},
    "rolled-back" => true
}
```

Reply:
Success even though entry does not exist:
```
[standalone@embedded /] /subsystem=elytron/jaas-realm=a:add(entry=e2,path=~/.java.login.config)
{"outcome" => "success"}
```

Reply:
Would we be better getting issues into WFWIP so we can triage them there?

Reply:
@darranl Just to be sure, do we want to check if the entry from the JAAS config file exists here (during definition of the realm)? Or when it is actually being used for credentials verification?

Reply:
We definitely should not be validating during Stage.MODEL as that stage can not guarantee the runtime environment, I suspect we also can't verify in Stage.RUNTIME as we still need to resolve the relative-to but it could be correct to validate as the service starts if we know the realm will never operate correctly due to a missing entry,

Reply:
@darranl Makes sense. The entry is required, so I added the check to the jaasRealmService's get() method then. As it can throw StartException it seems to be a good place. Thanks for explanation.

Reply:
Actually I made a mistake, sorry for confusion. I did not actually add this check. The reason is that LoginContext does not provide an API for me to find out if this entry exists before the LoginContext is being used for actual login. Currently, if the entry does not exist it will throw:

```
org.wildfly.security.auth.server.RealmUnavailableException: ELY01008: Failed to create login context
...
Caused by: javax.security.auth.login.LoginException: No LoginModules configured for NotConfiguredEntry
```

Btw it is possible to specify entry with name "other" to the JAAS config file. Java will use it (if it is defined) in cases where no other entry name matches.